### PR TITLE
Allow inspection and revocation of permissions

### PR DESCRIPTION
**Pull request by fd**
This patch allows deno scripts to revoke permissions granted by the user.  When the script, at a later stage, needs those permissions, a new prompt will be presented to the user.

The main use-case for this feature is to allow trusted _bootstrap_ code to run with raised privileges without granting all permissions to untrusted code following the bootstrap fase.

### CODE REVIEW

**Comment by zekth** [comment]
So one of the main goal for revocation of permissions is for example:

- Script is needing to load configuration from a file but do not need to read anything anymore
- deno runtime allow-read
- read the file
- revoke permission

Right?

I really like this PR. (but frustrated was working on it too :D )

**Comment by fd** [comment]
That's right. The only tricky part is that the _import order_ must deterministic and all _bootstrap_ code must be synchronous for this to work. Dynamic `import()` will eventually fix that too as it will allow us to import _unsafe code_ at after the first event-loop tick.

**Comment by afinch7** [comment]
Dynamic import wouldn't let us error before execution, but like any other import function It should give a reliable way of determining where the request comes from.

This one looks like it should work well. I can't say if it would get approved in the current state. Typically changes that add ops need to be pretty well justified to be accepted. 

I'm also not so sure how I feel about the revoke part, since I expect it to go unused in most cases. I think that separate allow once and allow always options for the permissions request prompt might be more useful.

File: js/permissions_test.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      assertEqual(perms[perm], perm === grant);
+    }
+
+    Deno.revokePermission(grant);
+
+    const revoked = Deno.permissions();
+    for (let perm of Object.keys(revoked)) {
+      assertEqual(revoked[perm], false);
+    }
+  });
```
Comment (ry) [inline js/permissions_test.ts]: Cool - interesting test

File: js/permissions.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+import * as msg from "gen/msg_generated";
+import * as flatbuffers from "./flatbuffers";
+import * as dispatch from "./dispatch";
+import { assert } from "./util";
+
+// Permissions as granted by the caller
+export type Permissions = {
```
Comment (ry) [inline js/permissions.ts]: Use jsdoc for the comment

File: js/permissions.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  env: boolean;
+  run: boolean;
+
+  // NOTE: Keep in sync with src/permissions.rs
+};
+
+export type Permission = keyof Permissions;
+
+// Inspect granted permissions for the current program.
+export function permissions(): Permissions {
```
Comment (ry) [inline js/permissions.ts]: Use jsdoc for the comment. Bonus points for a code example (ideally a working one that people can copy and paste)

File: js/permissions.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  const baseRes = dispatch.sendSync(...getReq())!;
+  assert(msg.Any.PermissionsRes === baseRes.innerType());
+  const res = new msg.PermissionsRes();
+  assert(baseRes.inner(res) != null);
+  // TypeScript cannot track assertion above, therefore not null assertion
+  return createPermissions(res);
+}
+
+// Revoke a permission. When the permission was already revoked nothing changes
+export function revokePermission(permission: Permission): void {
```
Comment (ry) [inline js/permissions.ts]: Jsdoc

File: js/permissions_test.ts (hunk 1 above)
Comment (zekth) [reply js/permissions_test.ts]: this can be refactored with
```typescript
for (let perm in revoked) {
```

Comment (afinch7) [reply js/permissions_test.ts]: `perm` could also be a `const`
```
for (const perm of of revoked) {
```

Comment (fd) [reply js/permissions_test.ts]: done, it no reads:

```ts
for (const perm in revoked) {
```

File: js/permissions.ts (hunk 2 above)
Comment (fd) [reply js/permissions.ts]: done

File: js/permissions.ts (hunk 3 above)
Comment (fd) [reply js/permissions.ts]: done

File: js/permissions.ts (hunk 4 above)
Comment (fd) [reply js/permissions.ts]: done

**Pull request merged**